Code/GraphMol/RascalMCES/RascalResult.cpp

```diff
@@ -0,0 +1,817 @@
+    applyMaxFragSep();
+  }
+  if (singleLargestFrag) {
+    largestFragOnly();
+  }
+}
+
+RascalResult::RascalResult(double tier1Sim, double tier2Sim)
+    : d_tier1Sim(tier1Sim), d_tier2Sim(tier2Sim) {}
+
+RascalResult::RascalResult(const RascalResult &other)
+    : d_bondMatches(other.d_bondMatches),
+      d_atomMatches(other.d_atomMatches),
+      d_smarts(other.d_smarts),
+      d_timedOut(other.d_timedOut),
+      d_tier1Sim(other.d_tier1Sim),
+      d_tier2Sim(other.d_tier2Sim),
+      d_numFrags(other.d_numFrags),
+      d_ringNonRingBondScore(other.d_ringNonRingBondScore),
+      d_atomMatchScore(other.d_atomMatchScore),
+      d_maxDeltaAtomAtomDist(other.d_maxDeltaAtomAtomDist),
+      d_largestFragSize(other.d_largestFragSize) {
+  if (other.d_mol1) {
+    d_mol1.reset(new ROMol(*other.d_mol1));
+  }
+  if (other.d_mol2) {
+    d_mol2.reset(new ROMol(*other.d_mol2));
+  }
+  if (other.d_mcesMol) {
+    d_mcesMol.reset(new ROMol(*other.d_mcesMol));
+  }
+}
+
+RascalResult &RascalResult::operator=(const RascalResult &other) {
+  if (this == &other) {
+    return *this;
+  }
+  d_bondMatches = other.d_bondMatches;
+  d_atomMatches = other.d_atomMatches;
+  d_smarts = other.d_smarts;
+  d_timedOut = other.d_timedOut;
+  d_numFrags = other.d_numFrags;
+  d_ringNonRingBondScore = other.d_ringNonRingBondScore;
+  d_atomMatchScore = other.d_atomMatchScore;
+  d_maxDeltaAtomAtomDist = other.d_maxDeltaAtomAtomDist;
+  d_largestFragSize = other.d_largestFragSize;
+  if (other.d_mol1) {
+    d_mol1.reset(new ROMol(*other.d_mol1));
+  }
+  if (other.d_mol2) {
+    d_mol2.reset(new ROMol(*other.d_mol2));
+  }
+  if (other.d_mcesMol) {
+    d_mcesMol.reset(new ROMol(*other.d_mcesMol));
+  }
+  return *this;
+}
+
+void RascalResult::largestFragOnly() { largestFragsOnly(1); }
+
+void RascalResult::largestFragsOnly(unsigned int numFrags) {
+  std::unique_ptr<RDKit::ROMol> mol1_frags(makeMolFrags(1));
+  // getMolFrags() returns boost::shared_ptr.  Ho-hum.
+  auto frags = RDKit::MolOps::getMolFrags(*mol1_frags, false);
+  if (numFrags < 1 || frags.size() < numFrags) {
+    return;
+  }
+  std::sort(frags.begin(), frags.end(),
+            [](const boost::shared_ptr<ROMol> &f1,
+               const boost::shared_ptr<ROMol> &f2) -> bool {
+              return f1->getNumAtoms() > f2->getNumAtoms();
+            });
+  frags.erase(frags.begin() + numFrags, frags.end());
+  rebuildFromFrags(frags);
+}
+
+void RascalResult::trimSmallFrags(unsigned int minFragSize) {
+  std::unique_ptr<RDKit::ROMol> mol1_frags(makeMolFrags(1));
+  // getMolFrags() returns boost::shared_ptr.  Ho-hum.
+  auto frags = RDKit::MolOps::getMolFrags(*mol1_frags, false);
+  for (auto &frag : frags) {
+    if (frag->getNumAtoms() < minFragSize) {
+      frag.reset();
+    }
+  }
+  frags.erase(std::remove_if(
+                  frags.begin(), frags.end(),
+                  [](const boost::shared_ptr<ROMol> &f) -> bool { return !f; }),
+              frags.end());
```

Comment:
maybe dumb question: couldn't you just do the check for the number of atoms in the `remove_if` and skip the first loop over frags on line 134?

Reply:
Doh!  Dumb code, not a dumb question!